packages/core/src/external/sftp/client.ts

```diff
@@ -0,0 +1,136 @@
+import SshSftpClient from "ssh2-sftp-client";
+import { Writable } from "stream";
+
+export interface SftpConfig {
+  host: string;
+  port: number;
+  username: string;
+  password: string;
+  privateKey: string;
+}
+
+export interface SftpClientImpl {
+  connect(): Promise<void>;
+  read(remotePath: string): Promise<Buffer>;
+  write(remotePath: string, content: Buffer): Promise<boolean>;
+  list(remotePath: string): Promise<string[]>;
+  exists(remotePath: string): Promise<boolean>;
+}
+
+export class SftpClient implements SftpClientImpl {
+  protected readonly client: SshSftpClient;
+  protected readonly host: string;
+  protected readonly port: number;
+  protected readonly username: string;
+  protected readonly password: string;
+  protected readonly privateKey: string;
+
+  constructor({ host, port, username, password, privateKey }: SftpConfig) {
+    this.client = new SshSftpClient();
+
+    this.host = host;
+    this.port = port;
+    this.username = username;
+    this.password = password;
+    this.privateKey = privateKey;
+  }
+
+  private async executeWithSshListeners<T, F extends (client: SshSftpClient) => Promise<T>>(
```

Comment:
Open question here - is this necessary? If the SSH connection drops before a transaction has been completed, we can always retry it and eventually achieve the correct result.